### PR TITLE
Exposing graph editor FRP outputs

### PR DESCRIPTION
**Pull request by wdanilo**
### Pull Request Description
- Exposes FRP outputs from the graph editor.
- Improves node selection.
- Adds basic shortcut docs.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md), [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: src/rust/ensogl/src/application/shortcut.rs [hunk 1]
```diff
@@ -15,6 +15,9 @@ use crate::frp;
 // === Registry ===
 // ================
 
+type RuleMap  = HashMap<KeyMask,Vec<WeakHandle>>;
+type ActionMap = HashMap<ActionType,RuleMap>;
```
Comment (farmaazon) [inline src/rust/ensogl/src/application/shortcut.rs:19]:
```suggestion
type RuleMap   = HashMap<KeyMask,Vec<WeakHandle>>;
type ActionMap = HashMap<ActionType,RuleMap>;
```

File: src/rust/ensogl/src/application/shortcut.rs [hunk 2]
```diff
@@ -25,46 +28,66 @@ use crate::frp;
...
+    }
+}
+
+impl Registry {
+    /// Constructor.
+    pub fn new(logger:&Logger, command_registry:&command::Registry) -> Self {
+        let model = RegistryModel::new(logger,command_registry);
+        frp::new_network! { network
+            eval model.keyboard.key_mask          ((m) model.process_action(ActionType::Press,m));
+            eval model.keyboard.previous_key_mask ((m) model.process_action(ActionType::Release,m));
```
Comment (farmaazon) [inline src/rust/ensogl/src/application/shortcut.rs:71]: what is meaning of m?
And this code looks a little hacky, that the events of previous_key_mask do the ActionType::Release. In fact the previous_key_mask should be updated on every event, not only on releases.

File: src/rust/ensogl/src/application/shortcut.rs [hunk 3]
```diff
@@ -128,52 +154,93 @@ impl Handle {
...
+    Press, Release
+}
+
+/// Keyboard action defined as `ActionType` and `KeyMask`, like "release key 'n'".
+#[derive(Clone,Copy,Debug)]
+#[allow(missing_docs)]
+pub struct Action {
+    pub tp       : ActionType,
+    pub key_mask : KeyMask,
+}
```
Comment (farmaazon) [inline src/rust/ensogl/src/application/shortcut.rs:194]: What is meaning of "releasing key_mask"? It is action of releasing all of keys in key_mask, or only a one of them? If I press Ctrl+Shift+S,and release, I can have three actions for releasing first S, then Shift, and Control?

File: src/rust/ensogl/src/application/shortcut.rs [hunk 4]
```diff
@@ -128,52 +154,93 @@ impl Handle {
...
+
+
+
 // ================
 // === Shortcut ===
 // ================
 
-/// A keyboard shortcut, a `keyboard::KeyMask` associated with a `Rule`.
+/// A keyboard shortcut, a `KeyMask` associated with a `Rule`.
```
Comment (farmaazon) [inline src/rust/ensogl/src/application/shortcut.rs:221]:
```suggestion
/// A keyboard shortcut, an `Action` associated with a `Rule`.
```

File: src/rust/lib/frp/src/macros.rs [hunk 5]
```diff
@@ -86,7 +86,7 @@ macro_rules! new_bridge_network {
 #[macro_export]
 macro_rules! _extend {
     ($trace:ident $network:ident $($ts:tt)*) => {
-        $crate::divide_on_terminator! { [[$crate::extend_lines] [$trace $network]] $($ts)* }
+        $crate::divide_on_terminator2! { [[$crate::extend_lines] [$trace $network]] $($ts)* }
```
Comment (farmaazon) [inline src/rust/lib/frp/src/macros.rs:174]: Too many 2 in macros names as for my taste. It makes the code more obscure that usually macros does.

File: src/rust/lib/frp/src/nodes.rs [hunk 6]
```diff
@@ -816,6 +970,242 @@ impl<Out:Data> stream::EventConsumer<Out> for OwnedMerge<Out> {
...
+}
+
+impl<T1> stream::InputBehaviors for IterData<T1> {
+    fn input_behaviors(&self) -> Vec<Link> {
+        vec![]
+    }
+}
+
+
+
```
Comment (farmaazon) [inline src/rust/lib/frp/src/nodes.rs:1208]: Many new nodes were introduced into FRP, but I don't see any of them tested.

File: src/rust/lib/graph-editor/src/lib.rs [hunk 7]
```diff
@@ -23,7 +23,7 @@
 #![warn(missing_copy_implementations)]
 #![warn(missing_debug_implementations)]
 
-#![recursion_limit="512"]
+#![recursion_limit="1024"]
```
Comment (farmaazon) [inline src/rust/lib/graph-editor/src/lib.rs:26]: image

File: src/rust/ensogl/src/application/shortcut.rs (hunk 2 above)
Comment (wdanilo) [reply src/rust/ensogl/src/application/shortcut.rs:71]: Oh, you are right!

File: src/rust/ensogl/src/application/shortcut.rs (hunk 3 above)
Comment (wdanilo) [reply src/rust/ensogl/src/application/shortcut.rs:194]: one of them. The key mask is the whole key binding, so releasing happens when it is no longer exactly in this state. Do you want better name or better docs?

Comment (farmaazon) [reply src/rust/ensogl/src/application/shortcut.rs:194]: Better docs will be nice

File: src/rust/ensogl/src/application/shortcut.rs (hunk 2 above)
Comment (wdanilo) [reply src/rust/ensogl/src/application/shortcut.rs:71]: I will leave fixme for next PR here.

Comment (wdanilo) [reply src/rust/ensogl/src/application/shortcut.rs:71]: Oh no, wait, it works fine :D

**Pull request merged**